### PR TITLE
Image slice request and response for async slicing

### LINKED ISSUE #5151: Add image slice function with request and response
# Description

The [accepted design for async slicing](https://napari.org/dev/naps/4-async-slicing.html?highlight=naps#detailed-description) defines a static method for slicing, with request and response types as the input and output.

The task here is to implement that method for the `Image` layer type.

See the [prototype PR for a crude reference implementation](https://github.com/napari/napari/pull/4905).

# Motivation

The slicing method is static so that it doesn't rely on any layer state directly and thus we don't need a lock to guard access, which could generally (across different layer types) result in blocking behavior. We could define this as a plain function instead of a static method on the layer, which may also help with code conflicts with `main`.

# Output

A PR that implements the image slice function with tests. I think there are existing tests that ensure `set_view_slice` does what it should, so we may want to adapt some of those instead of writing entirely new tests for this. Runs of existing benchmarks should show that this PR does not cause a major regression in performance.

# Questions

- Should we refactor `_set_view_slice` and `_get_slice` to share as much code as possible? I think the answer is yes, but that may make this task harder than desired.
- Should we also add the vispy image layer method that consumes the slice response here? Or in a separate PR?

### PR DESCRIPTION
**Pull request by andy-sweet**
# Description
This introduces image slice request and response types to be used by the new async slicing approach, as [described in NAP-4](https://napari.org/dev/naps/4-async-slicing.html). There are some small differences from the NAP that I hope/think are improvements and better communicate the intended approach here. The key difference is that the image request is now a callable that returns a response.

When we call `Image._make_slice_request`, we're combining the image and provided `Dims` to capture the input arguments to a function we can call later to get a slice response. This allows us to execute the request immediately on the calling thread (as is done here with some refactoring), but also allows us to execute the request asynchronously in the future too with worrying about calls to `Image.__setattr__` changing what something like `Image.data` refers to.

In an alternative async approach, we could transform the request class into a method of `Image` and acquire a lock that also guards mutation of the (e.g. `Image.data`). That reduces the number of classes and simplifies the code, but also introduces a lock that complicates the code and may also block unnecessarily sometimes.

One downside of this PR is that it adds some add extra complication to the already complicated image slicing logic, but this should be temporary while we need to continue to support the experimental async and traditional sync slicing approaches. I tried to add comments and docstrings to clarify where that temporary scaffolding is.

## Type of change
- [x] New feature (non-breaking change which adds functionality)

# References
Closes #5151 
Related to #4795 

# How has this been tested?
- [x] all existing tests pass with my change
- [x] benchmarks show no performance regressions

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

### CODE REVIEW

File: napari/layers/image/_slice.py
Comment (kephale) [inline napari/layers/image/_slice.py:130]: Is it useful for us to maintain our own transform implementation?

File: napari/layers/image/image.py
Comment (kephale) [inline napari/layers/image/image.py:750]: This is where the work is being triggered. This ended up being a nice surgical way of introducing request and response usage, now it will be easy to move the `execute()` call to a separate thread to prevent blocking of the main thread.

File: napari/layers/image/image.py
Comment (kephale) [inline napari/layers/image/image.py]: When I first skimmed through this implementation, I didn't realize what this function did because the name is a little confusing.

Comment (kephale) [reply napari/layers/image/image.py]: Just to clarify the confusion, the name made me think it was just a setter.

**Review by kephale — COMMENTED** [review]
I like these changes and only have an issue with the 1 function name.

File: napari/layers/image/image.py
Comment (andy-sweet) [reply napari/layers/image/image.py]: In the future, it might just be a setter. Or it might go away entirely. But I appreciate that right now it does a lot more (i.e. there are a lot of side-effects). Would `_update_slice_response` be better? Any other ideas?

Comment (kephale) [reply napari/layers/image/image.py]: `_update_slice_response` would be great!

Comment (andy-sweet) [reply napari/layers/image/image.py]: Done.

**Comment by andy-sweet** [comment]
Even though you might suspect otherwise, I believe [the macOS CI failures](https://github.com/napari/napari/actions/runs/3333433238/jobs/5515340670) are unrelated to this PR, as they are occurring on main too and likely due to a recent release of `pint` and a potentially flaky test.

**Comment by andy-sweet** [comment]
@jni : as discussed in the architecture working group, this should be ready to review now. I selectively ran some benchmarks locally and saw small improvements (and no regressions), but I'm currently running the full suite for a more complete picture.

**Comment by andy-sweet** [comment]
We need a better way to parse the benchmark run/comparison output, but I did try to dig through them to find a summary of performance changes.

Looks like some (maybe just one?) non-qt benchmark improved: https://github.com/napari/napari/actions/runs/3333481947/jobs/5515442677#step:7:5318

And there was no significant change in the qt benchmarks (even though the job failed): https://github.com/napari/napari/actions/runs/3333481947/jobs/5515442639#step:7:3601

File: napari/layers/image/_slice.py
Comment (Czaki) [inline napari/layers/image/_slice.py]:
```suggestion
    data : array like
```

File: napari/layers/image/_slice.py
Comment (Czaki) [inline napari/layers/image/_slice.py]:
```suggestion
    thumbnail: array like or none
```

**Review by Czaki — COMMENTED** [review]
in the docstring we should use type readable by a person, not exactly existing Python types. So I prefer to annotate `array like` instead of `Any`

File: napari/layers/image/_slice.py
Comment (andy-sweet) [reply napari/layers/image/_slice.py:130]: Do you mean our own transform implementation for slicing? Maybe, but I think it is a requirement that whatever we return should be easily convertible into a transform that napari can use.

**Comment by andy-sweet** [comment]
@brisvag : I swapped you in for @jni - let me know if you don't think you'll have time to review this over the next few days.

**Comment by brisvag** [comment]
I'll try to take a look soon!

**Review by kephale — APPROVED** [review]

**Comment by kephale** [comment]
Based on the Architecture WG discussion and the bug I ran into on the async prototype (https://github.com/napari/napari/issues/5297), I think we need to keep the automatic downsampling. Otherwise we're losing support for larger images.

**Comment by andy-sweet** [comment]
> Based on the Architecture WG discussion and the bug I ran into on the async prototype (#5297), I think we need to keep the automatic downsampling. Otherwise we're losing support for larger images.

Automatic downsampling currently occurs on the vispy side of things in [`VispyImageLayer._set_node_data`](https://github.com/napari/napari/blob/4ebbf1e58ac258a27775f4a314c3aa53418503da/napari/_vispy/layers/image.py#L131). This PR doesn't change that, so I think things should continue to work. With a simple integration (i.e. where that continues to happen), this means that downsampling will happen on the main thread, which could be a little choppy, but likely OK for now (ideally we might run it on the slicing thread and have the max texture sizes be inputs to slicing).

**Comment by kephale** [comment]
Got it, thanks!

**Comment by andy-sweet** [comment]
@brisvag : I made some updates after #5170 was merged to show how `LayerSlicer` can be used with the code here to actually do async slicing (though not actually in napari yet, since `LayerSlicer` is not integrated with the model yet). I have [a draft PR on my fork](https://github.com/andy-sweet/napari/pull/34) based on this PR's branch, but didn't merge it immediately in case you started reviewing this.

If you haven't started, let me know and I'll merge it so you can review everything all at once.

**Comment by brisvag** [comment]
Go ahead and merge, I have not had timeto look into this yet!

**Comment by andy-sweet** [comment]
> Go ahead and merge, I have not had timeto look into this yet!

Just adding a couple more things, and then will do.

File: napari/components/_tests/test_world_coordinates.py
Comment (andy-sweet) [inline napari/components/_tests/test_world_coordinates.py:107]: The refactor here means we can skip one call to `_slice_indices` by storing the result in a local variable, giving us a minor performance boost in some cases.

**Comment by andy-sweet** [comment]
@brisvag : I merged in the other changes, so this is really ready for review now.

**Comment by andy-sweet** [comment]
Here are the benchmark results for `time_set_view_slice` in the 2D and 3D image suites. Regular image slicing is pretty fast anyway, so the numbers are generally small, but we do half the runtime for 3D here (likely due to only calling `_slice_indices` once now). That performance improvement could be done outside of this PR, but it's nice small perk :)

```
(napari-dev) ➜  napari git:(main) asv run --python=same --bench "benchmark_image_layer.Image[23]DSuite.time_set_view_slice"
· Discovering benchmarks
· Running 2 total benchmarks (1 commits * 1 environments * 2 benchmarks)
[  0.00%] ·· Benchmarking existing-py_Users_asweet_software_miniconda3_envs_napari-dev_bin_python
[ 25.00%] ··· Running (benchmark_image_layer.Image2DSuite.time_set_view_slice--)..
[ 75.00%] ··· benchmark_image_layer.Image2DSuite.time_set_view_slice                                                                                                                                                                                                        ok
[ 75.00%] ··· ======== ============
               param1              
              -------- ------------
                 16     61.2±0.9μs 
                 32     67.2±0.7μs 
                 64      63.9±1μs  
                128      64.7±2μs  
                256      63.3±3μs  
                512      64.2±2μs  
                1024     63.5±2μs  
                2048     64.4±1μs  
                4096     64.0±2μs  
              ======== ============

[100.00%] ··· benchmark_image_layer.Image3DSuite.time_set_view_slice                                                                                                                                                                                                        ok
[100.00%] ··· ======== ============
               param1              
              -------- ------------
                 16      828±20μs  
                 32      798±30μs  
                 64      784±20μs  
                128      790±30μs  
                256      733±40μs  
                512      760±30μs  
                1024    1.21±0.2ms 
              ======== ============

(napari-dev) ➜  napari git:(main) git switch async-slice-image-request
Switched to branch 'async-slice-image-request'
(napari-dev) ➜  napari git:(async-slice-image-request) asv run --python=same --bench "benchmark_image_layer.Image[23]DSuite.time_set_view_slice"
· Discovering benchmarks
· Running 2 total benchmarks (1 commits * 1 environments * 2 benchmarks)
[  0.00%] ·· Benchmarking existing-py_Users_asweet_software_miniconda3_envs_napari-dev_bin_python
[ 25.00%] ··· Running (benchmark_image_layer.Image2DSuite.time_set_view_slice--)..
[ 75.00%] ··· benchmark_image_layer.Image2DSuite.time_set_view_slice                                                                                                                                                                                                        ok
[ 75.00%] ··· ======== ============
               param1              
              -------- ------------
                 16     60.0±0.9μs 
                 32      57.3±1μs  
                 64      57.0±1μs  
                128      59.1±3μs  
                256      57.1±2μs  
                512     59.1±0.9μs 
                1024     59.0±1μs  
                2048     58.5±1μs  
                4096     59.6±1μs  
              ======== ============

[100.00%] ··· benchmark_image_layer.Image3DSuite.time_set_view_slice                                                                                                                                                                                                        ok
[100.00%] ··· ======== ==========
               param1            
              -------- ----------
                 16     412±8μs  
                 32     410±10μs 
                 64     418±10μs 
                128     412±10μs 
                256     412±20μs 
                512     432±10μs 
                1024    514±6μs  
              ======== ==========
```

File: napari/layers/image/image.py
Comment (kcpevey) [inline napari/layers/image/image.py:769]: I'm a bit confused by breaking up `_make_slice_request` and `make_slice_requet_internal`. Why not just include these in a single method?

File: napari/layers/image/_slice.py
Comment (kcpevey) [inline napari/layers/image/_slice.py:77]: +1 on this. Adding this logic here really cleans up the layer itself and makes a clear separation of the slicing logic 👍

File: napari/components/_tests/test_layer_slicer.py
Comment (kcpevey) [inline napari/components/_tests/test_layer_slicer.py:286]: Do you intend on adding a lock to the layers which have async added (as in the FakeAsyncLayer)? If so, seem like it should go in this PR and be added as a test alongside this one (with the layer lock instead of the data lock)

File: napari/layers/image/image.py
Comment (andy-sweet) [reply napari/layers/image/image.py:769]: `_ImageBase._set_view_slice` doesn't have access to `Dims` so can't call `_make_slice_request`. I describe this as temporary scaffolding that should go away once `_set_view_slice` does (see the docstring of `_make_slice_request_internal`).

Alternatively, we could change the expected signature of `_make_slice_request` to accommodate the signature of `_slice_dims` (i.e. by passing through only some attributes of `Dims`), but I liked the conciseness of just passing the one argument.

File: napari/components/_tests/test_layer_slicer.py
Comment (andy-sweet) [reply napari/components/_tests/test_layer_slicer.py:286]: No. `FakeAsyncLayer` already has a lock. I did make an update in this PR to pass that lock through to the request, since we now need to acquire it in `__call__`.

But the only reason to add any of these locks is to inject something that allows to control the flow of execution on the slicing thread. So I'm reluctant to test them explicitly.

File: napari/layers/base/base.py
Comment (brisvag) [inline napari/layers/base/base.py:914]: Are there plans for where/how this check will happen once the layer becomes completely unaware of slicing?

**Review by brisvag — COMMENTED** [review]
Took a general tour of the code, all looks good to me! The details are a bit complex to navigate, not having followed closely the previous PR. Is there something specific that you'd like more focused feedback on?

File: napari/layers/base/base.py
Comment (andy-sweet) [reply napari/layers/base/base.py:914]: I haven't thought this through entirely, since we won't be removing slice state on the layer for the first of async slicing.

But a natural place feels like the `LayerSlicer` because that's the thing that deals with requests, so it could cache the last (or last few) request-response mappings.

The tricky thing about moving the slice state from the layer is that invalidating the cache becomes harder because state updates (e.g. a change in transforms) to the layer may now need to invalidate the cache in the `LayerSlicer`.

**Comment by andy-sweet** [comment]
> Took a general tour of the code, all looks good to me! The details are a bit complex to navigate, not having followed closely the previous PR. Is there something specific that you'd like more focused feedback on?

A lot of the core slicing implementation code is the same, but has been moved/refactored. So looking at the diffs is not very informative - I'd suggest looking at the code that was changed in the new files.

I guess the `LayerSlicer` and its tests are probably the most uniquely alien thing to you - the changes to that are fairly minimal and I also assigned @kcpevey to take a look at those as the author of the previous PR that introduced them.

So I think the best things to focus on are the following.

- The high level design changes in the image layer. The callable request type, response type, and `Layer._make_slice_request` and is one we intend to use for all layer types.
- The temporary/internal scaffolding code need to avoid breaking the existing slicing design.
- Can you think of anything that would improve the design here?
- Are the docs, examples, or tests that would help you understand the intended design?

Also see the PR description for some justification and reasoning.

File: napari/components/_layer_slicer.py
Comment (kcpevey) [inline napari/components/_layer_slicer.py:108]: I kind of want to see a test that would check both of these code paths to ensure the logic is working as expected. I just don't know how it would be implemented. I don't think the data lock is sufficient though, right? What do you think?

Comment (kcpevey) [reply napari/components/_layer_slicer.py:108]: on second thought, I think this should be tested at the base LayerSlicer level. I do have ideas for that which I'll include in my PR (https://github.com/napari/napari/pull/5299)

Comment (kcpevey) [reply napari/components/_layer_slicer.py:108]: My test checks both of these code paths, but it will need to be enhanced/modified once this PR with the `runtime_checkable` gets in.

**Comment by brisvag** [comment]
>    The high level design changes in the image layer. The callable request type, response type, and Layer._make_slice_request and is one we intend to use for all layer types.

I like the callable approach, I agree that it makes the break cleaner for later. One thing I'm wondering is: should *some* functionality at least be shared between each layer for the response object, via inheritance or otherwise? I assume so, but it's unclear to me at the moment what are the plans there.

>    The temporary/internal scaffolding code need to avoid breaking the existing slicing design.

Seems sensible to me. I has the same reaction as @kcpevey on the double method with an "internal" version, but I see you already thought about it so that's ok.

> Can you think of anything that would improve the design here?

I'm not familiar with the old async code, and that seems to create extra hassle at the moment. I'm sure this was discussed at lenght, but my feeling is that we should simply break that and fully embrace the new code already :P

>    Are the docs, examples, or tests that would help you understand the intended design?

The tests are a bit hard to follow (lots of setup) for someone who's not familiar with the code, but this is complex stuff, so I'm not sure there's a solution here ^^'

**Comment by andy-sweet** [comment]
> ne thing I'm wondering is: should _some_ functionality at least be shared between each layer for the response object, via inheritance or otherwise? I assume so, but it's unclear to me at the moment what are the plans there.

The intent was for the response type to really be a data/value type - i.e. an immutable store of values that can be consumed by whatever cares about slicing (e.g. vispy). I wasn't expecting it to have any methods or functionality.

Do you mean the request type?

Or do you mean that responses should share some attribute names and maybe types where possible? I'm not sure I see the value in that since most consumers want the layer-type specific response attributes anyway.

Also, in general are comfortable approving/rejecting this PR? Or should we look for someone else to take a look too?

**Review by brisvag — APPROVED** [review]
I meant the request, sorry :P Yes, I'm approving. I also had a second look through the tests, which I understand now.

**Comment by andy-sweet** [comment]
Thanks for the review @brisvag. I'll merge this after 24 hours unless there are any objections.

**Comment by kephale** [comment]
@andy-sweet , just a gentle nudge.

**Pull request merged**
